src/shared/commentService.ts

```diff
@@ -18,29 +12,47 @@ interface CommentServiceResponse<T> {
 export default class CommentService {
   private topicComments: firebase.firestore.CollectionReference;
   private firebase: firebase.app.App;
+  private likeService: LikeService;
 
   /**
    * Constructor
    * @param firebase the firebase app
    * @param topicId The id of the topic document to add/delete
    * comments on
    */
-  constructor(firebase: firebase.app.App, topicId: string) {
+  constructor(
+    firebase: firebase.app.App,
+    likeService: LikeService,
+    blogsCollectionName: string,
+    topicId: string,
+    commentsCollectionName: string
+  ) {
     this.firebase = firebase;
     this.topicComments = this.firebase
       .firestore()
-      .collection(BLOGS_COLLECTION)
+      .collection(blogsCollectionName)
       .doc(topicId)
-      .collection(COMMENTS_COLLECTION);
+      .collection(commentsCollectionName);
+
+    this.likeService = likeService;
   }
 
-  async getComments(): Promise<CommentServiceResponse<Comment[]>> {
+  async getComments(): Promise<ServiceResponse<Comment[]>> {
     let comments: Comment[];
     let error: Error;
 
     try {
       const query = await this.topicComments.get();
       comments = query.docs.map(doc => Comment.fromFirebaseDoc(doc));
+
+      // get the likes for each comment
+      await Promise.all(
```

Comment:
Seems like because of firebase we are stuck having two steps right? For example, there is no way to query firebase and get the comments with all the likes as well in one call?

Reply:
That's correct, and it's probably the biggest flaw of this implementation (I want to setup a configuration flag to enable/disable the entire like feature, and then move the like querying into a different comment service). Unfortunately, Firebase does not support querying a document and it's sub-collections in one query.

Originally, I had the likes as part of the comment in an array. However, this created a security flaw because the entire like would need to be updated in order to like the comment. Firebase security rules are not powerful enough to enable only part of the comment and like array to be updated.

Reply:
Makes sense. I figured Array would not work here as they are very limited and your feature is pretty robust.